### ISSUE TITLE
Problem with idle consumers keep committing their offset

### ISSUE THREAD

**Issue by jfrabaute** [bsm/sarama-cluster#124]
Hi,

We've seen a change in our services after we upgraded to the latest version of the sarama-cluster code, having this commit:

["Commit every time, even if not dirty."](https://github.com/bsm/sarama-cluster/commit/96927bc1831fb883c7903d3d48f5212041c9b1b5)

We see two behaviors that we didn't have previously and that we don't think should happen:

1. Filling the `"__consumer_offsets-XX.log"` files: As the offset is committed every X (default=250ms), this is logged by kafka, and the `.log` file is growing even when the consumer is idle:

```
{"timestamp":1493676453116,"name":"record_data","value":"\u0000\u0001\u0000\u0000\u0000\u0000\u0000\u0000\u0015�\u0000\u0000\u0000\u0000\u0001[�\r��\u0000\u0000\u0001[�4H�","offset":5545,"key":"\u0000\u0001\u0000\u0007testjfr\u0000\u0012__consumer_offsets\u0000\u0000\u0000(","topic":"__consumer_offsets","partition":40}
{"timestamp":1493676453154,"count":1,"name":"records_consumed","partitions":[{"topic":"__consumer_offsets","partition":40,"count":1,"minOffset":5545,"maxOffset":5545}]}
{"timestamp":1493676453155,"offsets":[{"topic":"__consumer_offsets","partition":40,"offset":5546}],"success":true,"name":"offsets_committed"}
```
When multiple consumers are connected to kafka, this can impact disk I/O.

2. Viewing "transient" consumers in the kafka outputs (using [kafka-manager](https://github.com/yahoo/kafka-manager) and kafka-* commands): What we call "transient" consumers are consumers that connect to a topic, consume messages (from newest) but never commit their offset. If they fail or disconnect, they reconnect and consume again from newest.
With the ["Commit every time, even if not dirty."](https://github.com/bsm/sarama-cluster/commit/96927bc1831fb883c7903d3d48f5212041c9b1b5) change, we now see those consumers displayed as consumers with offsets although they are not important for us, this info is not very useful and we don't want to commit any offset for them.

So, what is the rationale behind this ["Commit every time, even if not dirty."](https://github.com/bsm/sarama-cluster/commit/96927bc1831fb883c7903d3d48f5212041c9b1b5) change?
Is that something that could be reverted or modified?
Or do we miss some option to change the behavior of this feature?

Thank you.

**Comment by dim** [comment]
Hey, 

Thanks for the detailed report, the rationale is described here https://github.com/bsm/sarama-cluster/issues/116.

A quick question: what if you don't call `MarkOffset`? Do the `__consumer_offsets-*.log` files still grow? I would assume Kafka would only fill them with data if there is something new and for e.g. "transient consumers", you would not want to call `MarkOffset` anyway.

**Comment by dim** [comment]
I have created a new PR in #125, which force-pushes offsets only if Consumer.Offset.Retention setting is set. Please let me know if that's what you are looking for.

**Comment by jfrabaute** [comment]
Hi,

Thanks for the quick answer!

> Hey,
> 
> Thanks for the detailed report, the rationale is described here #116.

That's interesting.
The `__consumer_offsets` topic used by kafka to save the offset has `cleanup.policy = compact`. Doesn't it ensure the offset is not lost as the compaction policy will retain the last value for a specific key?
I've seen the doc here: [Log Compaction](https://kafka.apache.org/documentation/#compaction) and that's what I understand from it.

`Log compaction ensures that Kafka will always retain at least the last known value for each message key within the log of data for a single topic partition. It addresses use cases and scenarios such as restoring state after application crashes or system failure, or reloading caches after application restarts during operational maintenance.`

> 
> A quick question: what if you don't call MarkOffset? 

I don't call it when idle.
I call it only when I have processed messages (for **non** transient consumers), and I don't send any message on my test so I don't receive anything/don't call MarkOffset.
To reproduce the problem, here is the small code I'm using:
```go
package main

import (
	"fmt"
	"log"
	"os"
	"os/signal"
	"time"

	"github.com/Shopify/sarama"
	"github.com/bsm/sarama-cluster"
)

func main() {

	config := cluster.NewConfig()
	config.Consumer.Offsets.Initial = sarama.OffsetNewest
	config.Group.Session.Timeout = 10 * time.Second
	config.ClientID = "testconsumer"
	config.Consumer.Return.Errors = true
	config.Group.Return.Notifications = true

	brokers := []string{"localhost:9092"}
	topics := []string{"testidle"}
	consumer, err := cluster.NewConsumer(brokers, "testconsumergroup", topics, config)
	if err != nil {
		panic(err)
	}
	defer consumer.Close()

	// trap SIGINT to trigger a shutdown.
	signals := make(chan os.Signal, 1)
	signal.Notify(signals, os.Interrupt)

	// consume messages, watch errors and notifications
	for {
		select {
		case msg, more := <-consumer.Messages():
			if more {
				fmt.Fprintf(os.Stdout, "%s/%d/%d\t%s\t%s\n", msg.Topic, msg.Partition, msg.Offset, msg.Key, msg.Value)
				consumer.MarkOffset(msg, "") // mark message as processed
			}
		case err, more := <-consumer.Errors():
			if more {
				log.Printf("Error: %s\n", err.Error())
			}
		case ntf, more := <-consumer.Notifications():
			if more {
				log.Printf("Rebalanced: %+v\n", ntf)
			}
		case <-signals:
			return
		}
	}
}
```

I start zookeeper+kafka on my laptop (kafka with the config `auto.create.topics.enable = true`).
Here is the output of the go app:

```
$ go run consumer.go
2017/05/02 07:37:32 Rebalanced: &{Claimed:map[testidle:[0]] Released:map[] Current:map[testidle:[0]]}
```

And here is what's happening for the growing log file monitored with:

```sh
cd /tmp/kafka-logs/__consumer_offsets-49
watch -n 1 ls -la
```

I get this:

`-rw-r--r--   1 user  wheel     56633 May  2 07:47 00000000000000000000.log`
then 1 second later:
`-rw-r--r--   1 user  wheel     57021 May  2 07:47 00000000000000000000.log`

etc.

> Do the __consumer_offsets-*.log files still grow? 

Yes :disappointed:

> I would assume Kafka would only fill them with data if there is something new and for e.g. "transient consumers", you would not want to call MarkOffset anyway.

I thought so too, but it looks like kafka does not behave like this unfortunately. the broker seems to process all the requests and log them.

> I have created a new PR in #125, which force-pushes offsets only if Consumer.Offset.Retention setting is set. Please let me know if that's what you are looking for.

Thanks. This fixes the problem for "transient consumers". That seems like a good solution for this case.
But the growing logs problem is still there for consumers that really want to MarkOffset. When they are idle and they don't consume anything because the queue is not receiving any message, the `cmLoop` will commit the offset and the log file will grow.

Now, I'm wondering if #116 is really needed as the cleanup.policy is `compact` for the `__consumer_offsets` topic so it should ensure the latest offset is always present.

What do you understand from the kafka doc about this `compact` policy?

Thanks.

**Comment by jfrabaute** [comment]
More info about this offset retention policy.
While I was writing the answer, the kafka broker was running, doing nothing and I have those logs:

```
[2017-05-02 07:47:01,541] INFO [Group Metadata Manager on Broker 0]: Removed 0 expired offsets in 4 milliseconds. (kafka.coordinator.GroupMetadataManager)
[2017-05-02 07:57:01,517] INFO [Group Metadata Manager on Broker 0]: Removed 0 expired offsets in 1 milliseconds. (kafka.coordinator.GroupMetadataManager)
[2017-05-02 08:07:01,497] INFO [Group Metadata Manager on Broker 0]: Removed 0 expired offsets in 1 milliseconds. (kafka.coordinator.GroupMetadataManager)
[2017-05-02 08:17:01,481] INFO [Group Metadata Manager on Broker 0]: Removed 0 expired offsets in 1 milliseconds. (kafka.coordinator.GroupMetadataManager)
[2017-05-02 08:27:01,454] INFO [Group Metadata Manager on Broker 0]: Removed 0 expired offsets in 0 milliseconds. (kafka.coordinator.GroupMetadataManager)
```

Isn't the problem reported on #116 that the default offset retention for the is the same as the offsets.retention.minutes on the broker?

```
			// The retention duration for committed offsets. If zero, disabled
			// (in which case the `offsets.retention.minutes` option on the
			// broker will be used).  Kafka only supports precision up to
			// milliseconds; nanoseconds will be truncated. Requires Kafka
			// broker version 0.9.0 or later.
			// (default is 0: disabled).
			Retention time.Duration
```

If when creating the consumer, the `config.Consumer.Offsets.Retention` is changed to a high value, then kafka should keep this offset for a longer time. I'm not 100% sure, but not using this config param might be the cause of the problem for #116 .

**Comment by dim** [comment]
I think the problem of #116 is that retention client or server is reasonably short and offsets might expire if the consumer is not committing for a long period of time. Think of it this way:

1. you have a retention time of 1h
2. the consumer doesn't see any new messages overnight, hence there's nothing dirty and nothing will be committed to the broker
3. the broker expiration will trigger and your offset expires
4. the consumer is (accidentally) restarted and start consuming from the earliest offset again

I am sure commit #125 address both issues

**Comment by jfrabaute** [comment]
> I think the problem of #116 is that retention client or server is reasonably short and offsets might expire if the consumer is not committing for a long period of time. Think of it this way:
> 
> you have a retention time of 1h

Isn't it the problem? If the consumer does not want 1h (the default), it should change this in its config by doing:

```go
config.Consumer.Offsets.Retention = time.Hour * 168 // 1 week offset retention for this consumer
```
> the consumer doesn't see any new messages overnight, hence there's nothing dirty and nothing will be committed to the broker
> the broker expiration will trigger and your offset expires
> the consumer is (accidentally) restarted and start consuming from the earliest offset again
> I am sure commit #125 address both issues

I'll read more about #125 and see if that fixes the problem.

**Comment by jfrabaute** [comment]
For info: This [Kafka ticket](https://issues.apache.org/jira/browse/KAFKA-3806) talks about this problem.

**Comment by dim** [comment]
So I am more than happy to revert #117 and add a comment in the configuration. It seems as if this is more of a config problem than something that sarama-cluster should solve. Thanks for the links and the explanations, BTW.

@vanillechai is that something you agree with?

**Comment by jfrabaute** [comment]
> So I am more than happy to revert #117 and add a comment in the configuration. 

Thanks. That sounds good to me.

> It seems as if this is more of a config problem than something that sarama-cluster should solve. 

Yes, agreed.

> Thanks for the links and the explanations, BTW.

No problem.
I learned about this subtle config as well, and we changed our cluster config internally already to have the offset.retention == log.retention as the kafka ticket recommends.

Thanks for implementing and maintaining sarama-cluster!

**Comment by jfrabaute** [comment]
I finally found some time to send a pull request for this revert: #131

**Comment by dim** [comment]
This should be resolved by #125, I have released a new tag/version as well

**Comment by jfrabaute** [comment]
I tested with the latest version and I can confirm it behaves correctly now.

Thanks!
